src/compiler/checker.ts

```diff
@@ -12919,6 +12919,24 @@ namespace ts {
                 isUnitType(type);
         }
 
+        function isNullableDiscriminable(type: Type): boolean {
+            if (type.flags & TypeFlags.Union) {
+                let hasDiscriminant = false;
+                for (const current of (<UnionType>type).types) {
+                    if (current.flags & TypeFlags.Nullable) {
+                        hasDiscriminant = true;
+                        continue;
+                    }
+                    else if (current.flags & (TypeFlags.DisjointDomains | TypeFlags.Object)) {

```

Comment:
As a signpost for @ahejlsberg, I wasn't sure about this flag combination `(TypeFlags.DisjointDomains | TypeFlags.Object)`. I was mostly concerned about avoiding unions that had a generic parameter that *might* be instantiated to something *nullable*, but felt it was safer to be positive about what is accepted, rather than negative about what is not. I also considered whether it would be just sufficient to check the union has at least one nullable and at least one known non-nullable.